### PR TITLE
Change logging to use Serilog

### PR DESCRIPTION
**Pull request by rjmholt**
After having a few problems with logging and thread safety, I've replaced our hand-rolled logger with Serilog, and in such a way that only one file depends on it. Note that we also have to copy the Serilog DLLs into the module.

There are probably a couple of rough edges that still need patching up here, so please give it a try. Especially on .NET Framework platforms.

### CODE REVIEW

**Comment by tylerl0706** [comment]
Can you add the Serilog reference to the TPN?

https://github.com/PowerShell/PowerShellEditorServices/blob/master/Third%20Party%20Notices.txt

**Comment by rjmholt** [comment]
Looks like the logging tests are failing!

**Comment by rjmholt** [comment]
I should add a TPN for each of the Serilog bits I assume, since they are separate.

Also, is there a reason the Third Party Notice file has quotes in the name? It's named `'Third Party Notices.txt'` and not, e.g., `Third Party Notices.txt`.

**Comment by rjmholt** [comment]
Ok, I've been trying to get Serilog to log errors properly and I pretty much give up

**Comment by rjmholt** [comment]
There's a whole templating language and it's not exactly documented

**Comment by rjmholt** [comment]
I've tried destructuring, a template name, changing variable names...

**Comment by rjmholt** [comment]
Hmmm, actually it might work and we just weren't throwing exceptions when I thought we were

**Comment by rjmholt** [comment]
Nup wasn't working

**Comment by rjmholt** [comment]
Ok, it works now. Not perfectly, but I'm not sure how to indent the stack trace properly (since Serilog does it internally).

File: src/PowerShellEditorServices/Utility/Logging.cs [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+                    return LogEventLevel.Information;
+
+                case LogLevel.Warning:
+                    return LogEventLevel.Warning;
+
+                case LogLevel.Error:
+                    return LogEventLevel.Error;
+            }
+
+            throw new ArgumentException(String.Format("Unknown LogLevel: '{0}'", logLevel), nameof(logLevel));
```
Comment (rkeithhill) [inline src/PowerShellEditorServices/Utility/Logging.cs]: Could use C# string interpolation here `$"Unknown LogLevel: '{logLevel}'`.

File: test/PowerShellEditorServices.Test.Protocol/Message/MessageReaderWriterTests.cs [hunk 2]
```diff
@@ -20,12 +20,12 @@ public class MessageReaderWriterTests
...
         readonly int ExpectedMessageByteCount = Encoding.UTF8.GetByteCount(TestEventString);
 
-        private ILogger logger;
+        private IPsesLogger logger;
         private IMessageSerializer messageSerializer;
 
         public MessageReaderWriterTests()
         {
-            this.logger = new NullLogger();
+            this.logger = Logging.CreateLogger().Build();
```
Comment (rkeithhill) [inline test/PowerShellEditorServices.Test.Protocol/Message/MessageReaderWriterTests.cs]: So this is the equivalent of a Null logger?

Comment (rjmholt) [reply test/PowerShellEditorServices.Test.Protocol/Message/MessageReaderWriterTests.cs]: Yeah, it's not as explicit, but Serilog with no sinks does nothing. We could make a wrapper method maybe?

File: test/PowerShellEditorServices.Test/Console/ChoicePromptHandlerTests.cs [hunk 3]
```diff
@@ -96,7 +96,7 @@ internal class TestChoicePromptHandler : ChoicePromptHandler
 
         public int TimesPrompted { get; private set; }
 
-        public TestChoicePromptHandler() : base(new NullLogger())
+        public TestChoicePromptHandler() : base(Logging.CreateLogger().Build())
```
Comment (rkeithhill) [inline test/PowerShellEditorServices.Test/Console/ChoicePromptHandlerTests.cs]: Yeah, perhaps having a static CreateNullLogger method or something like that would be more clear/explicit

**Comment by rjmholt** [comment]
Serilog defines it's own `ILogger` type. I suppose I could have used namespaces to resolve that -- it didn't occur to me for some reason. I'm about to get rid of the interface actually and just have the class, since the build configures the class' internals and we're now using composition.

File: PowerShellEditorServices.build.ps1 [hunk 4]
```diff
@@ -177,13 +177,13 @@ task TestServer -If { !$script:IsUnix } {
     exec { & $script:dotnetExe xunit -configuration $Configuration -framework net452 -verbose -nobuild }
 }
 
-task TestProtocol -If { !$script:IsUnix} {
+task TestProtocol -If { !$script:IsUnix } {
```
[the remark this answers is not in the document]
Comment (rkeithhill) [reply PowerShellEditorServices.build.ps1]: Sorry, bad example.  There were some trailing spaces after a chained exception filter that got removed.  That's where the editorconfig file would have helped me where I use VS and it doesn't normally trim trailing whitespace.

**Comment by rkeithhill** [comment]
Less code churn is usually better unless there's a good reason for it.  :-)

File: src/PowerShellEditorServices/Analysis/AnalysisService.cs [hunk 5]
```diff
@@ -101,7 +101,7 @@ public string SettingsPath
         /// Creates an instance of the AnalysisService class.
         /// </summary>
         /// <param name="settingsPath">Path to a PSScriptAnalyzer settings file.</param>
-        /// <param name="logger">An ILogger implementation used for writing log messages.</param>
+        /// <param name="logger">A logger used for writing log messages.</param>
```
Comment (rkeithhill) [inline src/PowerShellEditorServices/Analysis/AnalysisService.cs]: Might as well revert all these comment changes.  That removes ~25 more files from the PR.  :-)

**Comment by rkeithhill** [comment]
BTW what does the log file look like with this change?

**Comment by rjmholt** [comment]
Logging looks like this at the moment:
```
2018-05-20 19:15:35.609 [NORMAL] /home/rob/Documents/Dev/Microsoft/PowerShellEditorServices/src/PowerShellEditorServices.Host/EditorServicesHost.cs: In 'StartLogging', line 166:
    PowerShell Editor Services Host v1.7.0.0 starting (pid 20583)...
    
      Host application details:
    
        Name:      Visual Studio Code Host
        ProfileId: Microsoft.VSCode
        Version:   1.7.0
    
2018-05-20 19:15:35.627 [NORMAL] /home/rob/Documents/Dev/Microsoft/PowerShellEditorServices/src/PowerShellEditorServices.Host/EditorServicesHost.cs: In 'StartLanguageService', line 198:
    Language service started, type = Tcp, endpoint = 17901
2018-05-20 19:15:35.634 [NORMAL] /home/rob/Documents/Dev/Microsoft/PowerShellEditorServices/src/PowerShellEditorServices.Host/EditorServicesHost.cs: In 'StartDebugService', line 268:
    Debug service started, type = Tcp, endpoint = 20769
2018-05-20 19:15:36.472 [NORMAL] /home/rob/Documents/Dev/Microsoft/PowerShellEditorServices/src/PowerShellEditorServices/Session/PowerShellContext.cs: In 'Initialize', line 220:
    PowerShell runtime version: 6.1.0, edition: Core
2018-05-20 19:15:37.196 [NORMAL] /home/rob/Documents/Dev/Microsoft/PowerShellEditorServices/src/PowerShellEditorServices.VSCode/ComponentRegistration.cs: In 'Register', line 36:
    PowerShell Editor Services VS Code module loaded.
2018-05-20 19:15:37.278 [ERROR] /home/rob/Documents/Dev/Microsoft/PowerShellEditorServices/src/PowerShellEditorServices.Protocol/MessageProtocol/MessageDispatcher.cs: In 'DispatchMessage', line 147:
    MessageDispatcher: No handler registered for Event type 'initialized'
2018-05-20 19:15:37.279 [ERROR] /home/rob/Documents/Dev/Microsoft/PowerShellEditorServices/src/PowerShellEditorServices.Protocol/MessageProtocol/MessageDispatcher.cs: In 'DispatchMessage', line 147:
    MessageDispatcher: No handler registered for Event type '$/setTraceNotification'
```

**Comment by rkeithhill** [comment]
I wonder if we should consider adding a `newline` after the log messages (except for the first one above)?  It might make visually scanning the log file a bit easier.  Then again, it could be that I'm just used to the spacing on the existing log file.  :-)

**Comment by rjmholt** [comment]
Actually I was thinking the same thing!

File: src/PowerShellEditorServices/Utility/Logging.cs [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                _filePaths.Add(filePath, logLevel);
+                return this;
+            }
+
+            /// <summary>
+            /// Configure the ILogger to send log messages to the console.
+            /// </summary>
+            /// <param name="logLevel">The minimum log level for console logging.</param>
+            /// <returns>the ILogger builder for reuse.</returns>
+            public Builder AddConsoleLogging(LogLevel? logLevel = null)
```
Comment (tylerl0706) [inline src/PowerShellEditorServices/Utility/Logging.cs]: Will this show up in the Integrated Terminal or in the VSCode logging view?

Comment (rjmholt) [reply src/PowerShellEditorServices/Utility/Logging.cs]: I'm not sure. Possibly the latter if VSCode redirects stdout to its logging view.

I put it in so you can start EditorServices by itself and see logs printed to the console you start it in. I haven't actually tested that that works though.

File: src/PowerShellEditorServices/Utility/Logging.cs [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+                {
+                    configuration = configuration.WriteTo.Async(a => a.File(logFile.Key,
+                        restrictedToMinimumLevel: ConvertLogLevel(logFile.Value ?? _logLevel),
+                        outputTemplate: "{Timestamp:yyyy-MM-dd HH:mm:ss.fff} {Message}{Newline}{Exception}{Newline}")
+                    );
+                }
+
+                return new PsesLogger(configuration.CreateLogger());
+            }
+
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Utility/Logging.cs]: Style nitpick:  Extra new line can be removed.

File: src/PowerShellEditorServices/Utility/Logging.cs [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        /// <summary>
+        /// A do-nothing logger that simply discards messages.
+        /// </summary>
+        public static ILogger NullLogger
+        {
+            get
+            {
+                return s_nullLogger ?? (s_nullLogger = CreateLogger().Build());
+            }
+        }
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Utility/Logging.cs]: Style nitpick: Add new line to separate members `NullLogger` and `s_nullLogger`

File: src/PowerShellEditorServices/Utility/PsesLogger.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        /// <param name="callerLineNumber">The line number in the file where the ILogger is being called.</param>
+        public void WriteException(
+            string errorMessage,
+            Exception exception,
+            [CallerMemberName] string callerName = null,
+            [CallerFilePath] string callerSourceFile = null,
+            [CallerLineNumber] int callerLineNumber = 0)
+        {
+            _logger.Error("[{Error:l}] {CallerSourceFile:l}: In method '{CallerName:l}', line {CallerLineNumber}:\n    {ErrorMessage:l}\n    {Exception:l}\n",
+                LogLevel.Error.ToString().ToUpper(), callerSourceFile, callerName, callerLineNumber, errorMessage, exception);
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Utility/PsesLogger.cs]: Should the result of `LogLevel.Error.ToString().ToUpper()` be a constant somewhere instead?  I'm assuming that `ToString().ToUpper()` will be called on each invocation regardless of log level.

File: src/PowerShellEditorServices/Utility/PsesLogger.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        private static string IndentMsg(string logMessage)
+        {
+            string[] msgLines = logMessage.Split('\n');
+
+            for (int i = 0; i < msgLines.Length; i++)
+            {
+                msgLines[i] = msgLines[i].Insert(0, "    ");
+            }
+
+            return String.Join("\n", msgLines)+"\n";
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Utility/PsesLogger.cs]: Is it feasible to use `StringBuilder` here to reduce some string allocations?  Something like

```csharp
var sb = new System.Text.StringBuilder(logMessage);
return sb.Replace("\n", "\n    ").ToString();
```

File: src/PowerShellEditorServices/Utility/PsesLogger.cs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        #region IDisposable Support
+        private bool disposedValue = false; // To detect redundant calls
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Utility/PsesLogger.cs]: Style nitpick: field name should be `_disposedValue` and comment should be above.

File: src/PowerShellEditorServices.Protocol/Server/LanguageServer.cs [hunk 12]
```diff
@@ -54,7 +54,7 @@ public IEditorOperations EditorOperations
         /// <param name="messageHandlers">An object that manages all of the message handlers</param>
         /// <param name="messageSender">The message sender</param>
         /// <param name="serverCompletedTask">A TaskCompletionSource<bool> that will be completed to stop the running process</param>
-        /// <param name="logger">The logger</param>
+        /// <param name="logger">the logger</param>
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply src/PowerShellEditorServices.Protocol/Server/LanguageServer.cs]: Not intentional — the result of changing all the logging doc lines and back again.

File: src/PowerShellEditorServices/Utility/PsesLogger.cs (hunk 11 above)
Comment (rjmholt) [reply src/PowerShellEditorServices/Utility/PsesLogger.cs]: Agreed, although this is what the "dispose pattern" snippet expands to.

**Comment by rjmholt** [comment]
Ok, I think I've finally understood Serilog's template string stuff. Because we do some indirection, we basically don't end up using a lot of it and do our own string processing, which I don't think can really be helped.

I've changed the newlines to use `Environment.NewLine`, which is what I suspect Serilog's `"{NewLine}"` string template expands to.

Our current log files look something like this:
```
2018-05-23 13:11:37.201 [NORMAL] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices.Host\EditorServicesHost.cs: In method 'StartLogging', line 166:
    PowerShell Editor Services Host v1.7.0.0 starting (pid 4324)...
    
      Host application details:
    
        Name:      Visual Studio Code Host
        ProfileId: Microsoft.VSCode
        Version:   1.7.0
        Arch:      64-bit
    
      Operating system details:
    
        Version: Microsoft Windows NT 10.0.16299.0
        Arch:    64-bit

2018-05-23 13:11:37.218 [NORMAL] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices.Host\EditorServicesHost.cs: In method 'StartLanguageService', line 198:
    Language service started, type = Tcp, endpoint = 18125

2018-05-23 13:11:37.222 [NORMAL] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices.Host\EditorServicesHost.cs: In method 'StartDebugService', line 268:
    Debug service started, type = Tcp, endpoint = 26849

2018-05-23 13:11:37.882 [NORMAL] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices\Session\PowerShellContext.cs: In method 'Initialize', line 220:
    PowerShell runtime version: 5.1.16299.431, edition: Desktop

2018-05-23 13:11:38.081 [ERROR] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices\Analysis\AnalysisService.cs: In method 'InvokePowerShell', line 445:
    The term 'Get-ScriptAnalyzerRule' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

2018-05-23 13:11:38.499 [NORMAL] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices.VSCode\ComponentRegistration.cs: In method 'Register', line 36:
    PowerShell Editor Services VS Code module loaded.

2018-05-23 13:11:38.656 [ERROR] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices.Protocol\MessageProtocol\MessageDispatcher.cs: In method 'DispatchMessage', line 152: Unhandled event
    System.Exception: VERY VERY BAD!!!
       at Microsoft.PowerShell.EditorServices.Protocol.MessageProtocol.MessageDispatcher.ThrowSomething()
       at Microsoft.PowerShell.EditorServices.Protocol.MessageProtocol.MessageDispatcher.<DispatchMessage>d__7.MoveNext()

2018-05-23 13:11:38.656 [ERROR] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices.Protocol\MessageProtocol\MessageDispatcher.cs: In method 'DispatchMessage', line 152: Unhandled event
    System.Exception: VERY VERY BAD!!!
       at Microsoft.PowerShell.EditorServices.Protocol.MessageProtocol.MessageDispatcher.ThrowSomething()
       at Microsoft.PowerShell.EditorServices.Protocol.MessageProtocol.MessageDispatcher.<DispatchMessage>d__7.MoveNext()

2018-05-23 13:11:39.459 [ERROR] C:\Users\roholt\Documents\Dev\PowerShellEditorServices\src\PowerShellEditorServices\Analysis\AnalysisService.cs: In method 'InvokePowerShell', line 445:
    The term 'Invoke-ScriptAnalyzer' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

```

**Comment by rjmholt** [comment]
Just to document:

Serilog has two levels of string template: the [output template](https://github.com/serilog/serilog/wiki/Formatting-Output) and the actual [event template](https://github.com/serilog/serilog/wiki/Writing-Log-Events).

The output template would be useful to us, since it has pre-defined settings for `{NewLine}`, (Log)`{Level}`, `{Timestamp}` and `{Exception}`. But these don't work so well in our case, e.g.:
- Our log levels are different from Serilogs, so we do so translation that Serilog can't understand.
- For `{Exception}` to work, [you need to log it without a template string](https://nblumhardt.com/2014/09/how-not-to-parameterize-serilog-events/), so we lose our log level and source metadata information.
- `{NewLine}` only exists in the output template, not the log template...

So instead we call `ToString()` on a few things and do some string building with indents.

But it's still very worth using Serilog, since it handles all the sinks and the asynchronous stuff very nicely.

**Comment by SeeminglyScience** [comment]
@rjmholt 
>  Because we do some indirection, we basically don't end up using a lot of it and do our own string processing

We might be able to do our own string processing and still let Serilog handle the when.  You can pass it a custom `Serilog.Formatting.ITextFormatter` implementation to the sink when configuring it.  There's also a way to [customize the formatting of specific objects](https://github.com/serilog/serilog/wiki/Formatting-Output#custom-text-formatters).

I wouldn't suggest holding up the PR for any of that though, this is already way better than what we currently have :)

**Comment by rjmholt** [comment]
Yeah, I actually implemented a custom format provider for our `LogLevel` enums at one stage because I didn't like calling `logLevel.ToString().ToUpper()`.

It took a surprising amount of code and playing around with the `$` and `@` destructuring stuff I still couldn't get it to work. That was where I submitted the series of comments above about not being able to make Serilog work. I understand it a bit better now I think, so it would be easier now. But still a fair amount of complexity and probably a slowdown basically for some indenting.

**Comment by rjmholt** [comment]
Anyway, unless there are any objections, I'll merge this at the end of the day (in about 8 hours).

**Pull request merged**